checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,363 @@
+      TreeUtils.getMethod(CreatesObligation.List.class, "value", 0, processingEnv);
+
+  /** The CreatesObligation.value field/element. */
+  final ExecutableElement createsObligationValueElement =
+      TreeUtils.getMethod(CreatesObligation.class, "value", 0, processingEnv);
+
+  /**
+   * Creates a MustCallAnnotatedTypeFactory.
+   *
+   * @param checker the checker associated with this type factory
+   */
+  public MustCallAnnotatedTypeFactory(final BaseTypeChecker checker) {
+    super(checker);
+    TOP = AnnotationBuilder.fromClass(elements, MustCallUnknown.class);
+    BOTTOM = createMustCall();
+    POLY = AnnotationBuilder.fromClass(elements, PolyMustCall.class);
+    addAliasedTypeAnnotation(InheritableMustCall.class, MustCall.class, true);
+    if (!checker.hasOption(MustCallChecker.NO_RESOURCE_ALIASES)) {
+      // In NO_RESOURCE_ALIASES mode, all @MustCallAlias annotations are ignored.
+      addAliasedTypeAnnotation(MustCallAlias.class, POLY);
+    }
+    this.postInit();
+  }
+
+  @Override
+  public void setRoot(@Nullable CompilationUnitTree root) {
+    super.setRoot(root);
+    elementsIssuedInconsistentMustCallSubtypeErrors.clear();
+    // TODO: this should probably be guarded by isSafeToClearSharedCFG from
+    // GenericAnnotatedTypeFactory,
+    // but this works here because we know the MCC is always the first subchecker that's sharing
+    // tempvars.
+    tempVars.clear();
+  }
+
+  @Override
+  protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+    // Explicitly name the qualifiers, in order to exclude @MustCallAlias.
+    return new LinkedHashSet<>(
+        Arrays.asList(MustCall.class, MustCallUnknown.class, PolyMustCall.class));
+  }
+
+  @Override
+  protected TreeAnnotator createTreeAnnotator() {
+    return new ListTreeAnnotator(super.createTreeAnnotator(), new MustCallTreeAnnotator(this));
+  }
+
+  @Override
+  protected void addComputedTypeAnnotations(Tree tree, AnnotatedTypeMirror type, boolean iUseFlow) {
+    super.addComputedTypeAnnotations(tree, type, iUseFlow);
+    // All primitives and boxed primitives are @MustCall({}). This code is needed to avoid primitive
+    // conversions, taking on the MustCall type of an object. For example, without this in this code
+    // b's type would be @MustCall("a"), which is nonsensical:
+    //
+    // @MustCall("a") Object obj; boolean b = obj == null;
+    if (TypesUtils.isPrimitiveOrBoxed(type.getUnderlyingType())) {
+      type.replaceAnnotation(BOTTOM);
+    }
+    if (isDeclaredInTryWithResources(TreeUtils.elementFromTree(tree))) {
+      type.replaceAnnotation(withoutClose(type.getAnnotationInHierarchy(TOP)));
+    }
+  }
+
+  @Override
+  public void addComputedTypeAnnotations(Element elt, AnnotatedTypeMirror type) {
+    super.addComputedTypeAnnotations(elt, type);
+    if (TypesUtils.isPrimitiveOrBoxed(type.getUnderlyingType())) {
+      type.replaceAnnotation(BOTTOM);
+    }
+    if (isDeclaredInTryWithResources(elt)) {
```

Comment:
> Or, what about writing the meta-annotation `@DefaultFor(RESOURCE_VARIABLE)` on `@MustCallUnknown`?

This code does not change the type of resource variables to top (i.e. `@MustCallUnknown`). Rather, it removes "close" from the type, if it is present, and leaves it unchanged otherwise. This operation can only result in a type *lower* in the hierarchy (e.g. `@MustCall("close")` becomes `@MustCall({})`, which is bottom, not top).

> maybe you can use a TreeAnnotator for resource variables

Good idea. I've done so.